academy/0.0-B9lab-Blockchains/5_crypto.md

```diff
@@ -0,0 +1,180 @@
+
+*  prove that you have a message without disclosing the content of the message, for instance:
+    * to prove you know your password
+    * to prove you previously wrote a message
+* rest assured the message was not altered
+* index your messages
+
+MD5 is such a function:
+
+```bash
+$ echo "The quick brown fox jumps over the lazy dog" | md5
+37c4b87edffc5d198ff5a185cee7ee09
+```
+
+On Linux, it is `md5sum`. Now let's introduce a typo:
+
+```bash
+$ echo "The quick brown fox jump over the lazy dog" | md5
+4ba496f4eec6ca17253cf8b7129e43be
+```
+
+Notice how the 2 hashes have nothing in common, other than the length, but the length is identical for all MD5 hashes so it reveals nothing about the input. 
+
+`MD5` is no longer considered a hard-to-crack hash function. Bitcoin uses `SHA-256`. Ethereum uses `Keccak-256`and `Keccak-512`.
+
+It is possible to index content by its hash, in essence creating a hashtable. If you have used IPFS or BitTorrent's magnet links, among others, then you have used a hashtable.
+
+### Digital Certificates
+
+Digital certificates are used (among other things) to prove an identity. They are given by a recognised Certification Authority(CA).  A widespread procedure is the public key certificate. It proves the ownership of a public key. We will describe below the standard X.509.
+
+The standard X.509 is defined by the Telecommunication Standardization Sector(ITU-T) of the International Telecommunication Union(ITU).[[1]](https://en.wikipedia.org/wiki/X.509) It offers format and semantics for public key certificates. [X.509](https://www.ietf.org/rfc/rfc5280.txt) is profiled in the formal language ASN.1. Common use cases are validation of documents and securing of communication. For an example, X.509 is used in TLS/SSL. Its origin is in the X.500 standard from year 1988. Since version 3, X.509 enables flexible topologies, like bridges and meshes. Invalid certificates are listed in certificate revocation lists. (CRLs) CRL is vulnerable against DOS attacks.
+
+A X.509 certificate contains information such as version number, serial number, signature algorithm, validity period, subject name, public key algorithm, subject public key, certificate signature algorithm, certificate signature and extensions. An extension has a type, a corresponding value and a critical flag. Non-critical extensions are only informational.
+
+#### Signature
+
+The concept of digital signatures is simple. If a given message is first hashed and then encrypted by a private key, one can verify the signature by decryption with corresponding public key. We need to hash the message to avoid the creation of signatures by mixing the messages and corresponding signatures. (See RSA chapter) This way, we know, that the sender has the private key to the given public key. 
+
+<HighlightBox type="reading">
+
+		* Zero-knowledge proof [https://en.wikipedia.org/wiki/Zero-knowledge_proof](https://en.wikipedia.org/wiki/Zero-knowledge_proof)
```

Comment:
Deleted this further reading as 0-knowledge-proof is not introduced once in the content.